src/mappings/utils/pools.ts

```diff
@@ -0,0 +1,168 @@
+        }
+
+        setPoolCurrencies(poolId, currencies);
+
+        if (currencies.length > index) {
+            return currencies[index];
+        }
+    } else {
+        throw Error("getStablePoolCurrencyByIndex: Unknown DexStablePoolsStorage version");
+    }
+
+    throw Error(`getStablePoolCurrencyByIndex: Unable to find currency in DexStablePoolsStorage for given poolId [${poolId}] and currency index [${index}]`);
+}
+
+function compareCurrencyType(currency0: Currency, currency1: Currency): number {
+    if (currency0.isTypeOf === currency1.isTypeOf) {
+        return 0;
+    }
+
+    const typeIndex0 = currencyTypeToIndexMap.get(currency0.isTypeOf);
+    const typeIndex1 = currencyTypeToIndexMap.get(currency1.isTypeOf);
+
+    if (typeIndex0 === undefined) {
+        throw Error(`Unable to find index for given currency type [${currency0.isTypeOf}]`);
+    }
+    if (typeIndex1 === undefined) {
+        throw Error(`Unable to find index for given currency type [${currency1.isTypeOf}]`);
+    }
+
+    return typeIndex0 - typeIndex1;
+}
+
+function currencyToIndex(currency: Currency): number {
+    switch(currency.isTypeOf) {
+        case "NativeToken":
+            const tokenIndex = nativeTokenToIndexMap.get(currency.token);
+            if (tokenIndex === undefined) {
+                throw Error(`currencyToIndex: Unknown or unhandled native token [${currency.token.toString()}]`);
+            }
+            return tokenIndex;
+        case "ForeignAsset":
+            return currency.asset;
+        case "LendToken":
+            return currency.lendTokenId;
+        case "StableLpToken":
+            return currency.poolId;
+        default:
+            throw Error(`currencyToIndex:  Unknown or unsupported currency type [${currency.isTypeOf}]`);
+    }
+}
+
+/**
+ * For sorting currencies.
+ * @param currency0 first currency
+ * @param currency1 second currency
+ * @returns A negative number if currency0 should be listed before currency1, 
+ *          a positive number if currency1 should be listed before currency0, 
+ *          otherwise returns 0
+ */
+function compareCurrencies(currency0: Currency, currency1: Currency): number {
+    const typeCompare = compareCurrencyType(currency0, currency1);
+    if (typeCompare != 0) {
+        return typeCompare;
+    }
+
+    const index0 = currencyToIndex(currency0);
+    const index1 = currencyToIndex(currency1);
+    return index0 - index1;
+}
+
+/**
+ * Calculate the standard pool's id given 2 currencies.
+ * This method will sort the currencies and return their ids/tickers in a specific order.
+ * 
+ * @param currency0 One currency
+ * @param currency1 The other currency
+ */
+export function inferGeneralPoolId(currency0: Currency, currency1: Currency): string {
+    let firstCurrencyString: string = currencyToString(currency0);
+    let secondCurrencyString: string = currencyToString(currency1);
+    
+    const order: number = compareCurrencies(currency0, currency1);
```

Comment:
I think I am little unsure here hmm, but why is it needed to sort the currencies here? Won't they be always returned from chain in the correct order? 

Is this way of indexing `currencyTypes` and sorting currencies mirroring how the order of currencies is set in parachain?

Reply:
The events provide the assets in the swap path in in/out order. After recombining multi-paths to pairs for assets and values, the order of assets can be anything, really. 
e.g. a swap path of `(KBTC, KSM, USDT)` would be broken down into [(KBTC, KSM) , (KSM, USDT)]`. In both cases, full path as well as split path, the currency order is only reflecting the direction of the swap.

Since the general pool swap event doesn't provide a pool ID in the event, I need to order the currencies so that `(KBTC, KSM)` and `(KSM, KBTC)` are identified as being part of the same pool as I aggregate the cumulative trading volumes.

Reply:
Didn't know event returns them in this way, thanks for clarifying! :raised_hands: